pyfixest/estimation/decomposition.py

```diff
@@ -0,0 +1,171 @@
+from typing import Any, Optional
+
+import numpy as np
+import pandas as pd
+from joblib import Parallel, delayed
+from tqdm import tqdm
+
+
+class LinearMediation:
+    """
+    Linear Mediation Model.
+
+    Initial implementation by Apoorva Lal at
+    https://gist.github.com/apoorvalal/e7dc9f3e52dcd9d51854b28b3e8a7ba4.
+    """
+
+    def __init__(
+        self,
+        agg: bool,
+        param: str,
+        coefnames: list[str],
+        nthreads: int = -1,
+        cluster_df: Optional[pd.Series] = None,
+    ):
+        self.cluster_dict: Optional[dict[Any, Any]]
```

Comment:
~17 lines of attaching objects to the class here (this is true for almost all estimator classes). Any downsides to making estimators [dataclasses](https://docs.python.org/3/library/dataclasses.html) ?

Reply:
So in principle, the idea would be to change to something like this: 

```python
from dataclasses import dataclass, field, InitVar
from typing import Any, Optional, List, Dict
import numpy as np
import pandas as pd

@dataclass
class GelbachDecomposition:
    """
    Linear Mediation Model.

    Initial implementation by Apoorva Lal at
    https://gist.github.com/apoorvalal/e7dc9f3e52dcd9d51854b28b3e8a7ba4.
    """
    
    agg: bool
    param: str
    coefnames: List[str]
    nthreads: int = -1
    cluster_df: Optional[pd.Series] = None
    
    # Define attributes initialized post-creation
    cluster_dict: Optional[Dict[Any, Any]] = field(init=False, default=None)
    unique_clusters: Optional[np.ndarray] = field(init=False, default=None)
    mask: np.ndarray = field(init=False)
    mediator_names: List[str] = field(init=False)
    X_dict: Dict[Any, Any] = field(init=False, default_factory=dict)
    X1_dict: Dict[Any, Any] = field(init=False, default_factory=dict)
    X2_dict: Dict[Any, Any] = field(init=False, default_factory=dict)
    Y_dict: Dict[Any, Any] = field(init=False, default_factory=dict)

    def __post_init__(self):
        # Set up mask to exclude 'param' from coefnames
        param_idx = self.coefnames.index(self.param)
        self.mask = np.ones(len(self.coefnames), dtype=bool)
        self.mask[param_idx] = False

        # Get mediator names excluding 'param'
        self.mediator_names = [name for name in self.coefnames if self.param not in name]

        # Handle clustering setup if cluster_df is provided
        if self.cluster_df is not None:
            self.unique_clusters = self.cluster_df.unique()
            self.cluster_dict = {
                cluster: self.cluster_df[self.cluster_df == cluster].index
                for cluster in self.unique_clusters
            }
        else:
            self.unique_clusters = None
            self.cluster_dict = None

        # Additional initialization for dictionaries can be added as needed

```
The main advantage is that there is a separation between `__init__` and processing of inputs in `__post__init__` which leads to cleaner code (which I like) but it does not really help to economize on lines of code, right? 

Generally I like it!

Reply:
Pushed something like the above.